### PR TITLE
Backend/change metro stations

### PR DESCRIPTION
**Pull request by vijaygupta18**
# 🚇 Multimodal Order Change Stops API - Complete Guide

## 📋 Overview
The `postMultimodalOrderChangeStops` API allows users to modify source and/or destination stations of journey legs in multimodal journeys. This document covers all possible scenarios, their responses, and what changes occur in the system.

---

## 🔗 API Endpoint
```
POST /v2/multimodal/{journeyId}/order/{legOrder}/changeStops
```

**Authentication**: Bearer token required  
**Content-Type**: application/json

---

## 📝 Request Structure
```json
{
  "journeyId": "string",
  "legOrder": "integer",
  "newSourceStation": {
    "stopCode": "string",
    "latLong": {
      "lat": "number",
      "lon": "number"
    }
  },
  "newDestinationStation": {
    "stopCode": "string", 
    "latLong": {
      "lat": "number",
      "lon": "number"
    }
  }
}
```

---

### **Scenario 1: Change Source Station Only**
**Request Body:**
```json
{
  "journeyId": "aa9ca558-7bb7-46de-b422-4cd865616c22",
  "legOrder": 1,
  "newSourceStation": {
    "stopCode": "4",
    "latLong": {
      "lat": 12.9760982,
      "lon": 77.6256022
    }
  }
}
```

**Response:**
```json
{
  "stationsChanged": true
}
```

**What Changes:**
- ✅ **Source Station**: Updated to new station (stopCode: "4")
- ✅ **Previous Walk Leg**: Distance, duration, and mode recalculated using OSRM
- ✅ **Previous Walk Leg**: May convert to Taxi if distance > `maxWalkDistance`
- ✅ **Metro Leg**: Source station details updated
- ✅ **Next Walk Leg**: Remains unchanged
- ✅ **Journey Leg**: New journey leg created, old one marked as deleted

---

### **Scenario 2: Change Destination Station Only**
**Request Body:**
```json
{
  "journeyId": "aa9ca558-7bb7-46de-b422-4cd865616c22",
  "legOrder": 1,
  "newDestinationStation": {
    "stopCode": "5",
    "latLong": {
      "lat": 12.916867,
      "lon": 77.6217365
    }
  }
}
```

**Response:**
```json
{
  "stationsChanged": true
}
```

**What Changes:**
- ✅ **Destination Station**: Updated to new station (stopCode: "5")
- ✅ **Previous Walk Leg**: Remains unchanged
- ✅ **Metro Leg**: Destination station details updated
- ✅ **Next Walk Leg**: Distance, duration, and mode recalculated using OSRM
- ✅ **Next Walk Leg**: May convert to Taxi if distance > `maxWalkDistance`
- ✅ **Journey Leg**: New journey leg created, old one marked as deleted

---

### **Scenario 3: Change Both Source and Destination Stations**
**Request Body:**
```json
{
  "journeyId": "aa9ca558-7bb7-46de-b422-4cd865616c22",
  "legOrder": 1,
  "newSourceStation": {
    "stopCode": "4",
    "latLong": {
      "lat": 12.9760982,
      "lon": 77.6256022
    }
  },
  "newDestinationStation": {
    "stopCode": "5",
    "latLong": {
      "lat": 12.916867,
      "lon": 77.6217365
    }
  }
}
```

**Response:**
```json
{
  "stationsChanged": true
}
```

**What Changes:**
- ✅ **Source Station**: Updated to new station (stopCode: "4")
- ✅ **Destination Station**: Updated to new station (stopCode: "5")
- ✅ **Previous Walk Leg**: Distance, duration, and mode recalculated using OSRM
- ✅ **Previous Walk Leg**: May convert to Taxi if distance > `maxWalkDistance`
- ✅ **Metro Leg**: Both source and destination station details updated
- ✅ **Next Walk Leg**: Distance, duration, and mode recalculated using OSRM
- ✅ **Next Walk Leg**: May convert to Taxi if distance > `maxWalkDistance`
- ✅ **Journey Leg**: New journey leg created, old one marked as deleted

---

### **Scenario 4: No Stations Provided (Error Case)**
**Request Body:**
```json
{
  "journeyId": "aa9ca558-7bb7-46de-b422-4cd865616c22",
  "legOrder": 1
}
```

**Response:**
```json
{
  "errorCode": "INVALID_STATION_CHANGE",
  "errorMessage": "Invalid station change for stop code : No stations provided for change",
  "errorPayload": null
}
```

**What Changes:**
- ❌ **Nothing Changes**: Request rejected
- ❌ **Error Returned**: INVALID_STATION_CHANGE error
- ❌ **HTTP Status**: 400 Bad Request

---

### **Scenario 5: Invalid Leg Order (Error Case)**
**Request Body:**
```json
{
  "journeyId": "aa9ca558-7bb7-46de-b422-4cd865616c22",
  "legOrder": 999,
  "newSourceStation": {
    "stopCode": "4",
    "latLong": {
      "lat": 12.9760982,
      "lon": 77.6256022
    }
  }
}
```

**Response:**
```json
{
  "errorCode": "INVALID_LEG_ORDER",
  "errorMessage": "Invalid leg order: 999",
  "errorPayload": null
}
```

**What Changes:**
- ❌ **Nothing Changes**: Request rejected
- ❌ **Error Returned**: INVALID_LEG_ORDER error
- ❌ **HTTP Status**: 400 Bad Request

---

### **Scenario 6: Invalid Station Code (Error Case)**
**Request Body:**
```json
{
  "journeyId": "aa9ca558-7bb7-46de-b422-4cd865616c22",
  "legOrder": 1,
  "newSourceStation": {
    "stopCode": "INVALID_STATION",
    "latLong": {
      "lat": 12.9760982,
      "lon": 77.6256022
    }
  }
}
```

**Response:**
```json
{
  "errorCode": "UNABLE_TO_CALL_NANDI_GET_STATIONS_BY_GTFS_ID_AND_STOP_CODE_API",
  "errorMessage": null,
  "errorPayload": null
}
```

**What Changes:**
- ❌ **Nothing Changes**: Request rejected
- ❌ **Error Returned**: Nandi API error (external service)
- ❌ **HTTP Status**: 500 Internal Server Error

---

### **1. Validation Phase**
- ✅ **Input Validation**: Check if at least one station is provided
- ✅ **Leg Validation**: Verify leg order exists in journey
- ✅ **Station Validation**: Validate station codes via Nandi API
- ✅ **Route Validation**: Check if valid metro route exists between stations

### **2. Walk Leg Updates**
- ✅ **OSRM Call**: Get actual walking distance and duration
- ✅ **Distance Check**: Compare with `riderConfig.maximumWalkDistance`
- ✅ **Mode Decision**: 
  - If distance ≤ `maxWalkDistance` → Keep as Walk
  - If distance > `maxWalkDistance` → Convert to Taxi
- ✅ **Data Update**: Update distance, duration, and mode in database

### **3. Metro Leg Updates**
- ✅ **Station Details**: Update source/destination station information
- ✅ **Route Details**: Update metro route information
- ✅ **Gate Information**: Update entrance/exit gate details

### **4. Database Changes**
- ✅ **New Journey Leg**: Create updated journey leg
- ✅ **Mark Old Leg**: Set old journey leg as deleted
- ✅ **Update Adjacent Legs**: Update previous and next walk legs

---

### **Success Response (All Scenarios)**
```json
{
  "stationsChanged": true
}
```

### **Error Response (All Error Scenarios)**
```json
{
  "errorCode": "ERROR_CODE_HERE",
  "errorMessage": "Human readable error message",
  "errorPayload": null
}
```

---

## 🚨 Error Codes and Meanings

| **Error Code** | **HTTP Status** | **When It Occurs** | **What It Means** |
|----------------|-----------------|-------------------|-------------------|
| `INVALID_STATION_CHANGE` | 400 | No stations provided | Request must include at least one station |
| `INVALID_LEG_ORDER` | 400 | Invalid leg sequence number | Leg order doesn't exist in journey |
| `STATION_NOT_FOUND` | 500 | Invalid station code | Station doesn't exist in system |
| `INVALID_STATION_DATA` | 400 | Missing station data | Required station information missing |
| `NO_VALID_METRO_ROUTE` | 400 | No route between stations | Metro can't connect the stations |
| `METRO_LEG_NOT_FOUND` | 400 | Metro leg issues | Problems with metro route structure |
| `OTP_SERVICE_UNAVAILABLE` | 503 | OTP service down | Transit planning service unavailable |
| `OSRM_FAILURE` | 500 | OSRM service error | Distance calculation service failed |

---

### **Core Functions**
1. **`postMultimodalOrderChangeStops`**: Main API endpoint
2. **`validateAndGetMetroLeg`**: Metro route validation
3. **`updateLegsWithGates`**: Walk leg updates with OSRM
4. **`getDistanceAndDuration`**: OSRM integration

### **OSRM Integration**
- **Purpose**: Calculate actual walking distances (not straight-line)
- **Benefit**: More accurate journey planning
- **Fallback**: Graceful degradation if OSRM fails

### **Smart Mode Conversion**
- **Walk → Taxi**: When distance > `maxWalkDistance`
- **Taxi → Walk**: When distance ≤ `maxWalkDistance`
- **Dynamic**: Real-time calculation based on actual routes

---

### **✅ Happy Path Tests**
- [ ] Change source station only
- [ ] Change destination station only  
- [ ] Change both stations
- [ ] Verify walk leg updates
- [ ] Verify mode conversion (Walk ↔ Taxi)

### **✅ Error Path Tests**
- [ ] No stations provided
- [ ] Invalid leg order
- [ ] Invalid station codes
- [ ] Network failures (OSRM/OTP)
- [ ] Authentication failures

### **✅ Edge Cases**
- [ ] Maximum distance boundaries
- [ ] Boundary coordinates
- [ ] Concurrent requests
- [ ] Large journey IDs
- [ ] Special characters in station codes

---

### **For Frontend Developers**
1. **Always validate**: Check both source and destination before sending
2. **Handle errors gracefully**: Show user-friendly error messages
3. **Update UI immediately**: Optimistically update UI, rollback on error
4. **Cache responses**: Store successful changes for offline reference

### **For Backend Developers**
1. **Validate early**: Check inputs before processing
2. **Handle failures gracefully**: OSRM/OTP failures shouldn't break the API
3. **Log everything**: Track all changes for debugging
4. **Use transactions**: Ensure database consistency

### **For QA Engineers**
1. **Test all scenarios**: Cover all possible combinations
2. **Test error conditions**: Verify error handling works
3. **Performance testing**: Check response times under load
4. **Integration testing**: Test with real OSRM/OTP services

## Summary by CodeRabbit

- **New Features**
  - Change source and/or destination stations for a metro leg in a multimodal journey via a new POST endpoint; returns confirmation when stations are changed.
  - Server now validates that at least one station is provided and performs route/gate-aware checks before applying changes.

- **Bug Fixes / Improvements**
  - Improved, user-facing error messages and HTTP codes for invalid station data, invalid leg order, routing failures, and service unavailability.
  - Better distance/duration checks to ensure route updates are valid.

### CODE REVIEW

File: Backend/app/rider-platform/rider-app/Main/spec/Storage/MultiModal.yaml [hunk 1]
```diff
@@ -297,6 +297,10 @@ JourneyLeg:
       kvFunction: updateWithKV
       params: [estimatedMinFare, estimatedMaxFare]
       where: legSearchId
+    updateIsDeleted:
```
Comment (hkmangla) [inline Backend/app/rider-platform/rider-app/Main/spec/Storage/MultiModal.yaml]: remove this.

**Review by hkmangla — DISMISSED** [review]

File: Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/MultimodalConfirm.hs [hunk 2]
```diff
@@ -1035,3 +1047,240 @@
...
+      legOrder
+      (Nothing, newLeg, Nothing)
+      journey.fromLocation journey.toLocation
+      journey.merchantId reqJourneyLeg.merchantOperatingCityId
+      journeyId
+      (Id journey.searchRequestId)
+      riderConfig.maximumWalkDistance Nothing mbGates
+  QJourneyLegMapping.updateIsDeleted True reqJourneyLeg.id
+  QJourneyLegExtra.create newJourneyLeg
+
```
[the remark this answers is not in the document]
Comment (vijaygupta18) [reply Backend/app/rider-platform/rider-app/Main/src/Domain/Action/UI/MultimodalConfirm.hs:1088]: jab nhi aata to jyda bolaa mat kr
